drivers/video/video_esp32_dvp.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (c) 2024 espros photonics Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_cam
+
+#include <soc/gdma_channel.h>
+#include <soc/lcd_cam_reg.h>
+#include <soc/lcd_cam_struct.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/dma/dma_esp32.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/video.h>
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(esp32_cam, LOG_LEVEL_INF);
+
+#define VIDEO_ESP32_F160M               160000000UL
+#define VIDEO_ESP32_DMA_BUFFER_MAX_SIZE 4095
+
+enum video_esp32_cam_clk_sel_values {
+	VIDEO_ESP32_CAM_CLK_SEL_NONE = 0,
+	VIDEO_ESP32_CAM_CLK_SEL_XTAL = 1,
+	VIDEO_ESP32_CAM_CLK_SEL_PLL_DIV2 = 2,
+	VIDEO_ESP32_CAM_CLK_SEL_PLL_F160M = 3,
+};
+
+struct video_esp32_config {
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *sensor_dev;
+	const struct device *clock_dev;
+	const clock_control_subsys_t clock_subsys;
+	const struct device *dma_dev;
+	uint8_t rx_dma_channel;
+	uint32_t clock_mclk;
+	int irq_source;
+	int irq_priority;
+	uint8_t enable_16_bit;
+	uint8_t invert_de;
+	uint8_t invert_byte_order;
+	uint8_t invert_bit_order;
+	uint8_t invert_pclk;
+	uint8_t invert_hsync;
+	uint8_t invert_vsync;
+};
+
+struct video_esp32_data {
+	const struct video_esp32_config *config;
+	struct video_format video_format;
+	struct video_buffer *active_vbuf;
+	bool is_streaming;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	uint8_t dma_buffer[VIDEO_ESP32_DMA_BUFFER_MAX_SIZE];
+	struct dma_block_config dma_blocks[CONFIG_DMA_ESP32_DESCRIPTOR_NUM];
+};
+
+struct video_esp32_dma_data {
+	struct k_fifo *fifo_out;
+	struct video_buffer *vbuf;
+};
+
+static int video_esp32_reload_dma(struct video_esp32_data *data);
+
+void video_esp32_dma_rx_done(const struct device *dev, void *user_data, uint32_t channel,
+			     int status)
+{
+	struct video_esp32_data *data = user_data;
+	int ret = 0;
+
+	if (status == DMA_STATUS_BLOCK) {
+		LOG_DBG("received block");
+		return;
+	}
+
+	if (status != DMA_STATUS_COMPLETE) {
+		LOG_ERR("DMA error: %d", status);
+		return;
+	}
+
+	if (data->active_vbuf == NULL) {
+		LOG_ERR("No video buffer available. Enque some buffers first.");
+		return;
+	}
+
+	k_fifo_put(&data->fifo_out, data->active_vbuf);
+	data->active_vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
```

Comment:
Interesting way to do it! With two FIFO, this is now completely asynchronous...

Reply:
Thanks. I think I got the inspiration from `video_stm32_dcmi` driver :-)